### ISSUE TITLE
The 'updated' filter does not consider comments on the diff of PRs

### ISSUE THREAD

**Issue by damithc** [HubTurbo/HubTurbo#654]
Looks like it does not consider the time stamp of comments added to the diff of PRs

**Comment by damithc** [comment]
Some other updates seem to be missing too. For example, some comments added about 7 hours ago are missing from the `updated:10` filter.

**Comment by damithc** [comment]
The missing updates may be due to running out of API quota.

**Comment by acruis** [comment]
Comments on diffs are under the category of [Review Comments](https://developer.github.com/v3/pulls/comments/), which are different from [Issue Comments](https://developer.github.com/v3/issues/comments/). There are also [Commit Comments] (https://developer.github.com/v3/repos/comments/), which are also not retrieved when we trigger an API call for Issue Comments. What this means is that the program will consume extra API calls and take more time until the updates arrive.

**Comment by damithc** [comment]
No choice. Have to retrieve them as well. :-)

**Comment by damithc** [comment]
Another type of comments we want to capture are comments on commits.

**Comment by ndt93** [comment]
There is a quite problematic issue arises when I'm trying to implement this.

When there is an updated filter, all issues will first be filtered using the issues' `updatedAt` time. Then the metadata will be obtained for the filtered issues. Problem is that adding a review comment or commit comment does not change an issue's `updatedAt` time. So issues which only have review comments in the valid window won't be in the filtered list in the first place.

Anyone has any suggestion on this?

**Comment by ndt93** [comment]
A solution is to treat `PullRequests` separately from `Issues` since the review comments affects `PullRequests`' updatedAt time. But from what I can imagine, this implies really sweeping changes over the whole application...

**Comment by dariusf** [comment]
What if we made TurboIssue.getUpdatedAt() transparently take the later of the issue's and PR's updated times? Would that work?

**Comment by ndt93** [comment]
The problem is getting all the PRs' `updatedAt`. Right now our mechanisms for getting new/updated issues revolve around the fact that everything is treated as issues. Treating PRs separately means that we have to replicate that whole mechanism for PRs since they need different requests for GitHubAPI and also different datastructure.

**Comment by ndt93** [comment]
I'm thinking about getting the updated PRs then inject the updatedTime into Issues. But we still probably need to store the ETags, request for updated PRs in the `UpdateService` and all those stuffs that we do for Issues.

**Comment by dariusf** [comment]
We do need to store an ETag to detect when the PR changes. When we get an Issue from GitHub, does it come with the updated PullRequest object? I'm under the impression that we can get the new update time at that point, as part of the current flow for updating issues.

**Comment by ndt93** [comment]
The GitHub API request for getting updated issues uses the `since` parameter which is compared against `updated_at` so the problem already starts from there. Also the `pull_request` object that comes with an issue contains only links to the actual PR so we will need extra 1 API call for every issue in that case.

**Comment by dariusf** [comment]
I see... Could we get the PRs separately as part of the metadata process? This would be where the second round of requests would go out.

**Comment by ndt93** [comment]
Hmm, I'm not entirely sure about this but seems like it's more consistent as part of the first round of updates since we would have to store some persistent data about these PRs while the Metadata is transient. Also, from the current mechanism, the metadata is requested only for issues already filtered by updatedAt time.

**Comment by ndt93** [comment]
Do you think this is complex enough to warrant a new issue?

**Comment by dariusf** [comment]
With the new design, I think the metadata phase can be extended to get any kind of extra data before presentation. Either way can work I think. I was thinking of it as being similar to how we fetch comments. If we get it as part of the issue, we would do all the ETag checks for the PR together with those for the issue?

Sure, we can break it up further.

**Comment by ndt93** [comment]
A potential problem I think of is that the `updatedAt` field is not only used for the `updated` filter, and it's one of the most important field that should be synced with GitHub as soon as possible and kept persistent, otherwise future features that rely on it can be caught off guard. And to guarantee consistency, every issues update must be accompanied by corresponding PRs update.

The metadata phase is designed to get extra data for only a few issues filtered by `updatedAt`, but there is a circular dependency here since we need the PRs' `updatedAt` data before we can filter. So the flow is: {Update Issues -> Update PRs -> Inject PR's `updatedAt` to Issues}  -> Filter by `updatedAt` -> [Get Metadata for selected Issues]

I think this can be done in this issue since I already integrate the review/commit comments. But this might not make it for the milestone.

**Comment by ndt93** [comment]
Another big problem crops up, GitHub API does not provide a convenient `since` parameters for pull requests so we can't get updated pull requests convenient like for issues. In that case we have to override PageIterator to use ETags for each page so we don't have to get all pull requests every time.

If there is no other solutions for this problem then this is definitely getting too big. I will write tests and merge the review comments integration first then create another issue for pull requests update.

**Comment by dariusf** [comment]
Okay, sure. What's involved with the review comments integration? Do you need the PRs to get them?

**Comment by ndt93** [comment]
No need the PRs directly but it needs the correct `updatedAt` field to behave correctly in all cases. So the integration is done but it doesn't feel fulfilling enough to be merged :/. Should I just do it here and merge all later for the next milestone after 3.8?

**Comment by dariusf** [comment]
Alright, no worries.
